diff --git a/airflow/models/dag.py b/airflow/models/dag.py
index 3fb306afc9..1bde3b88d5 100644
--- a/airflow/models/dag.py
+++ b/airflow/models/dag.py
@@ -2328,15 +2328,17 @@ class DAG(LoggingMixin):
         else:
             matched_tasks = [t for t in self.tasks if t.task_id in task_ids_or_regex]
 
-        also_include: list[Operator] = []
+        also_include: set[Operator] = set()
         for t in matched_tasks:
             if include_downstream:
-                also_include.extend(t.get_flat_relatives(upstream=False))
+                downstreams_to_add = t.get_flat_relatives(upstream=False)
+                also_include.update(downstreams_to_add)
             if include_upstream:
-                also_include.extend(t.get_flat_relatives(upstream=True))
+                upstreams_to_add = t.get_flat_relatives(upstream=True)
+                also_include.update(upstreams_to_add)
             else:
                 upstream_setup_tasks = t.get_flat_relatives(upstream=True, setup_only=True)
-                also_include.extend(upstream_setup_tasks)
+                also_include.update(upstream_setup_tasks)
 
         direct_upstreams: list[Operator] = []
         if include_direct_upstream:
